contracts/SafeProtocolMediator.sol

```diff
@@ -0,0 +1,168 @@
+            SafeProtocolAction memory safeProtocolAction = transaction.actions[i];
+            (bool isActionSuccessful, bytes memory resultData) = safe.execTransactionFromModuleReturnData(
+                safeProtocolAction.to,
+                safeProtocolAction.value,
+                safeProtocolAction.data,
+                0
+            );
+
+            // Need to revisit the approach below. If some actions fail, the transaction stiil succeeds.
+            // With current approach, even if one action fails, `data` will be empty bytes even for successful
+            // actions.
+            if (!isActionSuccessful) {
+                revert ActionExecutionFailed(address(safe), transaction.metaHash, i);
+            } else {
+                emit ActionExecuted(address(safe), transaction.metaHash, i);
+                data[i] = resultData;
+            }
+        }
+    }
+
+    /**
+     * @notice TODO
+     * @param safe TODO
+     * @param rootAccess TODO
+     * @return data TODO
+     */
+    function executeRootAccess(
+        ISafe safe,
+        SafeRootAccess calldata rootAccess
+    ) external override onlyEnabledModule(safe) returns (bytes memory data) {
+        SafeProtocolAction memory safeProtocolAction = rootAccess.action;
+        // TODO: Check for re-entrancy attacks
+        // TODO: Validate metahash
+
+        // Re-confirm if this check if needed and correct.
+        if (!ISafeProtocolModule(msg.sender).requiresRootAccess()) {
+            revert ModuleRequiresRootAccess(msg.sender);
+        }
+
+        if (!enabledComponents[address(safe)][msg.sender].rootAddressGranted) {
+            // TODO: Need new error type?
+            revert ModuleRequiresRootAccess(msg.sender);
+        }
+        bool success;
+        (success, data) = safe.execTransactionFromModuleReturnData(
+            safeProtocolAction.to,
+            safeProtocolAction.value,
+            safeProtocolAction.data,
+            1
+        );
+        if (success) {
+            emit RootAccessActionExecuted(address(safe), rootAccess.metaHash);
+        } else {
+            revert RootAccessActionExecutionFailed(address(safe), rootAccess.metaHash);
+        }
+    }
+
+    /**
+     * @notice Called by a Safe to enable a module on a Safe. To be called by a safe.
+     * @param module ISafeProtocolModule A module that has to be enabled
+     * @param allowRootAccess Bool indicating whether root access to be allowed.
+     */
+    function enableModule(ISafeProtocolModule module, bool allowRootAccess) external {
+        // TODO: Check if module is a valid address and implements valid interface.
+        //       Validate if caller is a Safe.
+        //       Should it be allowed to enable a module twice with different allowRootAccess flag?
+        if (allowRootAccess != module.requiresRootAccess()) {
+            revert ModuleAccessMismatch(address(module), module.requiresRootAccess(), allowRootAccess);
```

Comment:
While it is only for the revert case, but we make multiple calls to an external contract (`module`) which is quite expensive compared to remembering the state.